server/src/main/java/io/druid/segment/loading/LocalDataSegmentPusher.java

```diff
@@ -87,19 +92,37 @@ public DataSegment push(File dataSegmentFile, DataSegment segment) throws IOExce
       );
     }
 
+    File tmpOutDir = new File(baseStorageDir, storageDir + "." + UUID.randomUUID().toString());
+    long size = compressSegment(dataSegmentFile, tmpOutDir);
+
+    DataSegment dataSegment = createDescriptorFile(
+      segment.withLoadSpec(makeLoadSpec(new File(outDir, "index.zip")))
+             .withSize(size)
+             .withBinaryVersion(SegmentUtils.getVersionFromDir(dataSegmentFile)),
+      tmpOutDir
+    );
+
+    // moving the temporary directory to the final destination, once success the potentially concurrent push operations
+    // will be failed and will read the descriptor.json created by current push operation directly
+    try {
+      java.nio.file.Files.move(tmpOutDir.toPath(), outDir.toPath());
+    }
+    catch (FileAlreadyExistsException e) {
+      log.warn("Push destination directory[%s] exists, ignore this message if replication is configured.");
+      dataSegment = jsonMapper.readValue(new File(outDir, "descriptor.json"), DataSegment.class);
```

Comment:
I really prefer the temp files being kept in the same directory as configured for deep storage, for two reasons:

- For atomic rename of A -> B to work, A and B must be on the same filesystem. A good way to guarantee that is to have them be in the same directory. The java tmp dir is a particularly problematic choice since it's often on a tmpfs.
- Having them in deep storage actually makes them easier to clean up, I think, since operators just need to search through deep storage for files older than a certain date with names matching a certain pattern -- instead of having to track two different directories.